content/en/blog/2024/l7-policy-with-opa/index.md

```diff
@@ -0,0 +1,293 @@
+  "ok_response": {
+      "headers": [],
+      "headers_to_remove": [],
+      "dynamic_metadata": {...},
+      "response_headers_to_add": [],
+      "query_parameters_to_set": [],
+      "query_parameters_to_remove": []
+    },
+  "dynamic_metadata": {...}
+}
+{{< /text >}}
+
+This means that based on the response from the Authz server, the envoy can add or remove headers, query parameters, and even change the response status. So that OPA can do the same as it is described in the [OPA's documentation](https://www.openpolicyagent.org/docs/latest/envoy-primer/#example-policy-with-additional-controls).
+
+## Testing
+
+Let's test the simple usage (authorization) and then let's create a more advanced rule to show how we can use OPA to modify the request and response.
+
+Deploy an app to run curl commands to the Httpbin sample application:
+
+{{< text bash >}}
+$ kubectl -n my-app run --image=curlimages/curl curly -- /bin/sleep 100d
+{{< /text >}}
+
+Apply the first rego rule and restart the OPA deployment:
+
+{{< text bash >}}
+$ kubectl apply -f opa-example-1-configmap.yaml
+
+$ kubectl rollout restart deployment -n opa
+{{< /text >}}
+
+The simple scenario is to allow requests if they contain the header `x-force-authorized` with the value `enabled` or `true`. If the header is not present or has a different value, the request will be denied.
+
+There are multiple ways to create the rego rule. In this case, we created two different rules. Executed in order, the first one which satisfies all the conditions will be the one that will be used.
+
+### Simple rule
+
+The following request will return `403`:
+
+{{< text bash >}}
+$ kubectl exec -n my-app curly -c curly  -- curl -s -w "\nhttp_code=%{http_code}" httpbin/get
+{{< /text >}}
+
+The following request will return `200` and the body:
+
+{{< text bash >}}
+$ kubectl exec -n my-app curly -c curly  -- curl -s -w "\nhttp_code=%{http_code}" httpbin/get -H "x-force-authorized: enabled"
+{{< /text >}}
+
+### Advanced manipulations
+
+Now the more advanced rule. Apply the second rego rule and restart the OPA deployment:
+
+{{< text bash >}}
+$ kubectl apply -f opa-example-2-configmap.yaml
+
+$ kubectl rollout restart deployment -n opa
+{{< /text >}}
+
+In that rule, you can see:
+
+{{< text plain >}}
+myrule["allowed"] := allow # Notice that `allowed` is mandatory when returning an object, like here `myrule`
+myrule["headers"] := headers
+myrule["response_headers_to_add"] := response_headers_to_add
+myrule["request_headers_to_remove"] := request_headers_to_remove
+myrule["body"] := body
+myrule["http_status"] := status_code
+{{< /text >}}
+
+And those are the values that will be returned to the envoy proxy. The envoy will use those values to modify the request and response.
+
+Notice that `allowed is required when returning a JSON object instead of only true/false. This can be found in the documentation [here](https://www.openpolicyagent.org/docs/latest/envoy-primer/#output-document).
```

Comment:
```suggestion
Notice that `allowed` is required when returning a JSON object instead of only true/false. This can be found in the OPA documentation [here](https://www.openpolicyagent.org/docs/latest/envoy-primer/#output-document).
```